src/client.rs

```diff
@@ -0,0 +1,162 @@
+use cosmian_crypto_core::{
+    reexport::rand_core::SeedableRng, Aes256Gcm, CryptoCoreError, CsRng, Dem,
+    FixedSizeCBytes, Instantiable, Nonce, RandomFixedSizeCBytes, SymmetricKey,
+};
+use rand::{Rng, RngCore};
+use std::io::{Error, ErrorKind};
+
+use crate::{btree::DataItem, oram::BUCKET_SIZE};
+
+pub struct ClientORAM {
+    pub stash: Vec<DataItem>,
+    csprng: CsRng,
+    key: SymmetricKey<{ Aes256Gcm::KEY_LENGTH }>,
+}
+
+impl ClientORAM {
+    pub fn new() -> ClientORAM {
+        let mut csprng = CsRng::from_entropy();
+        let key = SymmetricKey::new(&mut csprng);
+
+        ClientORAM {
+            // Empty stash at initialization.
+            stash: Vec::new(),
+            csprng,
+            key,
+        }
+    }
+
+    pub fn gen_key(&mut self) {
+        self.key = SymmetricKey::new(&mut self.csprng);
+    }
+
+    pub fn generate_dummies(
+        &mut self,
+        nb_dummies: usize,
+        block_size: usize,
+    ) -> Result<Vec<DataItem>, Error> {
+        if !(nb_dummies + 1).is_power_of_two() {
+            return Err(Error::new(
+                ErrorKind::InvalidInput,
+                "Number of blocks shall be power of 2 minus one".to_string(),
+            ));
+        }
+
+        // Number of leaves is 2^(l-1) if number of blocks is 2^l - 1.
+        let nb_leaves = (nb_dummies + 1) / 2;
+        let mut dummies = Vec::new();
```

Comment:
Allocate the correct size here:
```suggestion
        let mut dummies = Vec::with_capacity(nb_dummies);
```

Reply:
Shouldn't it be
```
let mut dummies = Vec::with_capacity(nb_dummies * BUCKET_SIZE);
```
?

Reply:
Nope, your definition of `nb_dummies` is wrong. It actually counts the nodes (the size of `dummies` that you are declaring should be `nb_dummies`, otherwise this is obfuscating and you should rename one or the other).

Reply:
Commit aa8979d36c2b5943f9e657417a65ca8e58b92eca